src/jwt/sign.ts

```diff
@@ -1,149 +1,113 @@
    *
+   * @param protectedHeader JWS Protected Header.
    */
-  toSign(): string {
-    if (!this._protectedHeader || !this._protectedHeader.alg) {
-      throw new JWSInvalid(
-        'Missing signature algorithm. Specify the `alg` field'
-      );
-    }
-    const protectedHeader = encodeBase64(JSON.stringify(this._protectedHeader));
-    const payload = encodeBase64(JSON.stringify(this._payload));
-    if (payload && protectedHeader) {
-      return `${protectedHeader}.${payload}`;
-    } else {
-      throw new JWSInvalid();
-    }
+  setProtectedHeader(protectedHeader: Omit<JWSHeaderParameters, 'alg'>) {
+    this._protectedHeader = protectedHeader;
+    return this;
   }
 
   /**
-   * Append signature to unsigned JWT.
-   * For an ECDSA signature it is required that this is in ASN.1/DER encoded format.
-   * The same format used by the TEE. Conversion to JWS is handled automatically.
+   * Return a JWT without signature (`header.payload`).
    *
-   * @param jwtWithoutSignature
-   * @param signature
    */
-  static async appendSignature(
-    jwtWithoutSignature: string,
-    signature: string
-  ): Promise<string> {
-    if (
-      typeof jwtWithoutSignature !== 'string' ||
-      typeof signature !== 'string'
-    ) {
-      throw new JWTInvalid('JWS must be a string');
-    }
-    if (signature === '') throw new JWSInvalid('Invalid signature');
-
-    const jwtDecoded = SignJWT.decodeJwtWithoutSignature(jwtWithoutSignature);
-    const alg = jwtDecoded.header.alg;
-    const kty = getKtyFromAlg(alg);
-
-    if (kty === 'EC') {
-      const encodedJws = await derToJose(signature, alg);
-      return `${jwtWithoutSignature}.${encodedJws}`;
-    } else {
-      const encodedJws = removePadding(signature);
-      return `${jwtWithoutSignature}.${encodedJws}`;
-    }
+  async unsigned(): Promise<string> {
+    const { alg } = await this.getSelectedSigningAlgorithm();
+
+    return [{ ...this._protectedHeader, alg }, this._payload]
+      .map((e) => JSON.stringify(e))
+      .map(encodeBase64)
+      .join('.');
+  }
+
+  private async getSelectedSigningAlgorithm(): Promise<{
+    kty: string;
+    alg: SupportedAlgorithm;
+  }> {
+    const publicKey = await this.crypto.getPublicKey();
+    const alg = getAlgFromKey(publicKey);
+
+    return { kty: publicKey.kty, alg };
   }
 
   /**
-   * Decodes a JWT without signature
+   * Return a signed JWT.
    *
-   * @param jwtWithoutSignature JWT to sign that needs to be decoded.
    */
-  static decodeJwtWithoutSignature(jwtWithoutSignature: string): {
-    payload: JWTPayload;
-    header: CompactJWSHeaderParameters;
-  } {
-    if (typeof jwtWithoutSignature !== 'string') {
-      throw new JWTInvalid('JWT must be a string');
-    }
-    const {
-      0: encodedHeader,
-      1: encodedPayload,
-      length,
-    } = jwtWithoutSignature.split('.');
+  async signed(): Promise<string> {
```

Comment:
what do you think about renaming this function to `sign()` like [that](https://github.com/panva/jose/blob/main/docs/classes/jwt_sign.SignJWT.md)

Reply:
> what do you think about renaming this function to sign()

See no evil, we can do it. How do we rename `unsigned()`? (btw, does `unsigned()` make sense? I left it because of the older implementation, but I think we're not gonna need it anymore)

Reply:
> pass the keytag to the sign function

The `SignJWT` class knows nothing about keytags, or any other method to store/retrieve keys. It's actually the `CryptoContext` that hides such complexity.

The intended usage is:
* we create a `CryptoContext` instance which is bound to a specific key pair 
  * (that is: to a specific key tag)
* such context exposes methods such as `getSignature()` and `getPublicKeys()` that require no informations from the user
  * (that is, it's the context that known which key pair is bound to)

`SignJWT` cannot make mistakes when handling keys, as it's done by the context.
Moreover, a `CryptoContext` can work with the Trusted Execution Environment, as well as plain-text JWKs, or whatever.

Reply:
ok I agree. Then I would just say rename `signed()` to `sign()` and delete `unsigned()`. However, with the new version the implementation will have to change so continuing to keep `unsigned` makes no sense.

Reply:
Done in https://github.com/pagopa/io-react-native-jwt/pull/21/commits/905abadf30855df86bdc76a17ed87e10b7e192a7